contrib/coredns/policy/dns_test.go

```diff
@@ -300,6 +332,144 @@ func TestSetRedirectQueryAnswer(t *testing.T) {
 			";; QUESTION SECTION:\n;"+
 			"redirect.example.com.\tIN\t A\n",
 	)
+
+	m = testutil.MakeTestDNSMsg("example.com", dns.TypeA, dns.ClassINET)
+	w = testutil.NewTestAddressedNonwriter("192.0.2.1")
+
+	rc, err = p.setRedirectQueryAnswer(context.TODO(), w, m, "A=127.0.0.1;AAAA=23ef:3546::8732;TXT=Ghguyw7g7yiug7;CNAME=google.com")
+	if err != nil {
+		t.Error(err)
+	}
+	testutil.AssertDNSMessage(t, "setRedirectQueryAnswer(192.0.2.53)", rc, m, dns.RcodeSuccess,
+		";; opcode: QUERY, status: NOERROR, id: 0\n"+
+			";; flags:; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0\n\n"+
+			";; QUESTION SECTION:\n"+
+			";example.com.\tIN\t A\n\n"+
+			";; ANSWER SECTION:\n"+
+			"example.com.\t0\tIN\tA\t127.0.0.1\n",
+	)
+
+	m = testutil.MakeTestDNSMsg("example.com", dns.TypeAAAA, dns.ClassINET)
+	w = testutil.NewTestAddressedNonwriter("192.0.2.1")
+
+	rc, err = p.setRedirectQueryAnswer(context.TODO(), w, m, "A=127.0.0.1;AAAA=23ef:3546::8732;TXT=Ghguyw7g7yiug7;CNAME=google.com")
+	if err != nil {
+		t.Error(err)
+	}
+	testutil.AssertDNSMessage(t, "setRedirectQueryAnswer(2001:db8::53)", rc, m, dns.RcodeSuccess,
+		";; opcode: QUERY, status: NOERROR, id: 0\n"+
+			";; flags:; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0\n\n"+
+			";; QUESTION SECTION:\n"+
+			";example.com.\tIN\t AAAA\n\n"+
+			";; ANSWER SECTION:\n"+
+			"example.com.\t0\tIN\tAAAA\t23ef:3546::8732\n",
+	)
+
+	m = testutil.MakeTestDNSMsg("example.com", dns.TypeTXT, dns.ClassINET)
+	w = testutil.NewTestAddressedNonwriter("192.0.2.1")
+
+	rc, err = p.setRedirectQueryAnswer(context.TODO(), w, m, "A=127.0.0.1;AAAA=23ef:3546::8732;TXT=Ghguyw7g7yiug7;CNAME=google.com")
+	if err != nil {
+		t.Error(err)
+	}
+	testutil.AssertDNSMessage(t, "setRedirectQueryAnswer(2001:db8::53)", rc, m, dns.RcodeSuccess,
+		";; opcode: QUERY, status: NOERROR, id: 0\n"+
+			";; flags:; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0\n\n"+
+			";; QUESTION SECTION:\n"+
+			";example.com.\tIN\t TXT\n\n"+
+			";; ANSWER SECTION:\n"+
+			"example.com.\t0\tIN\tTXT\t\"Ghguyw7g7yiug7\"\n",
+	)
+
+	m = testutil.MakeTestDNSMsg("example.com", dns.TypeA, dns.ClassINET)
+	w = testutil.NewTestAddressedNonwriter("192.0.2.1")
+
+	rc, err = p.setRedirectQueryAnswer(context.TODO(), w, m, "AAAA=23ef:3546::8732;TXT=Ghguyw7g7yiug7;CNAME=google.com")
+	if err != errTest {
+		t.Error("Expected to be attempted to resolve by CNAME and get testError, got", err)
+	}
+	// mocked upstream response here, so no answer expected
+	testutil.AssertDNSMessage(t, "setRedirectQueryAnswer(2001:db8::53)", rc, m, dns.RcodeServerFailure,
+		";; opcode: QUERY, status: NOERROR, id: 0\n"+
+			";; flags:; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 0\n\n"+
+			";; QUESTION SECTION:\n"+
+			";example.com.\tIN\t A\n",
+	)
```

Comment:
why no answer? Should be CNAME and A records

Reply:
Yes, updated them. I had overlooked the mockplugin behavior a bit.